mongo/integration/collection_test.go

```diff
@@ -1154,6 +1155,124 @@ func TestCollection(t *testing.T) {
+
+			mt.ClearEvents()
+			res, err = mt.Coll.BulkWrite(mtest.Background, deleteModels)
+			assert.Nil(mt, err, "BulkWrite error: %v", err)
+			assert.Equal(mt, int64(numDocs), res.DeletedCount, "expected %v deleted documents, got %v", numDocs, res.DeletedCount)
+			mt.FilterStartedEvents(func(evt *event.CommandStartedEvent) bool {
+				return evt.CommandName == "delete"
+			})
+			// MaxWriteBatchSize changed between 3.4 and 3.6, so there isn't a given number of batches that this will be split into
+			deletes := len(mt.GetAllStartedEvents())
+			assert.True(mt, deletes > 1, "expected multiple batches, got %v", deletes)
+		})
+		mt.Run("update with batches", func(mt *mtest.T) {
+			// TODO(GODRIVER-425): remove this as part a larger project to
+			// refactor integration and other longrunning tasks.
+			// if os.Getenv("EVR_TASK_ID") == "" {
+			// 	mt.Skip("skipping long running integration test outside of evergreen")
+			// }
+
+			var models []mongo.WriteModel
+			numModels := 100050
+
+			// it's significantly faster to upsert one model and modify the rest than to upsert all of them
+			for i := 0; i < numModels-1; i++ {
+				update := bson.D{
+					{"$set", bson.D{
+						{"a", int32(i + 1)},
+						{"b", int32(i * 2)},
+						{"c", int32(i * 3)},
+					}},
+				}
+				model := mongo.NewUpdateOneModel().
+					SetFilter(bson.D{{"a", int32(i)}}).
+					SetUpdate(update).SetUpsert(true)
+				models = append(models, model)
+			}
+			// add one last upsert
+			models = append(models, mongo.NewUpdateOneModel().
+				SetFilter(bson.D{{"x", int32(1)}}).
+				SetUpdate(bson.D{{"$set", bson.D{{"x", int32(1)}}}}).
+				SetUpsert(true),
+			)
+
+			mt.ClearEvents()
+			res, err := mt.Coll.BulkWrite(mtest.Background, models)
+			assert.Nil(mt, err, "BulkWrite error: %v", err)
+
+			mt.FilterStartedEvents(func(evt *event.CommandStartedEvent) bool {
+				return evt.CommandName == "update"
+			})
+			// MaxWriteBatchSize changed between 3.4 and 3.6, so there isn't a given number of batches that this will be split into
+			updates := len(mt.GetAllStartedEvents())
+			assert.True(mt, updates > 1, "expected multiple batches, got %v", updates)
+
+			assert.Equal(mt, int64(numModels-2), res.ModifiedCount, "expected %v modified documents, got %v", numModels-2, res.ModifiedCount)
+			assert.Equal(mt, int64(numModels-2), res.MatchedCount, "expected %v matched documents, got %v", numModels-2, res.ModifiedCount)
+			assert.Equal(mt, int64(2), res.UpsertedCount, "expected %v upserted documents, got %v", 2, res.UpsertedCount)
+			assert.Equal(mt, 2, len(res.UpsertedIDs), "expected %v upserted ids, got %v", 2, len(res.UpsertedIDs))
+
+			id1, ok := res.UpsertedIDs[0]
+			assert.True(mt, ok, "expected id at key 0")
+			id2, ok := res.UpsertedIDs[int64(numModels-1)]
+			assert.True(mt, ok, "expected id at key %v", numModels-1)
+
+			doc, err := mt.Coll.FindOne(mtest.Background, bson.D{{"_id", id1}}).DecodeBytes()
```

Comment:
Two ideas to clean this up a bit. I don't have a preference between them:

1. Use the `Int32OK` variants to avoid doing a separate `LookupErr` call: `a, ok := doc.Lookup("a").Int32OK()`.

2. Use `compareDocs` instead:

```
expected := {"a": int32(numModels-1), "b": int32((numModels-2)*2), "c": int32((numModels-2)*3)}
compareDocs(expected, doc)
```

You'd have to build `expected` either using `bson.D` followed by `bson.Marshal` or using `bsoncore` directly (but not the `DocumentBuilder` type because that can't be backported to the 1.4 branch).

Reply:
done